endpoints/s3/src/completemultipartupload.cpp

```diff
@@ -404,27 +432,48 @@ void irods::s3::actions::handle_completemultipartupload(
 
 	// wait until all threads are complete
 	std::unique_lock<std::mutex> lk(cv_mutex);
-	cv.wait(lk, [&upload_status_object, max_part_number, func = __FUNCTION__]() {
+	cv.wait(lk, [&upload_status_object, max_part_number, func = __func__]() {
 		logging::debug("{}: wait: task_done_counter is {}", func, upload_status_object.task_done_counter);
 		return upload_status_object.task_done_counter == max_part_number;
 	});
 
-	d.close();
+	// close the object and delete the entry in the replica_token_and_number_map
+	if (part_shmem::replica_token_and_number_map.find(upload_id) != part_shmem::replica_token_and_number_map.end()) {
+		// Read all of the shared pointers in the tuple to make sure they are destructed in
+		// the order we require. std::tuple does not guarantee order of destruction.
+		auto conn_ptr = std::get<2>(part_shmem::replica_token_and_number_map[upload_id]);
+		auto transport_ptr = std::get<3>(part_shmem::replica_token_and_number_map[upload_id]);
```

Comment:
I was troubleshooting an occasional error due to the order of deletion of these things when going through the chunk encoding flow.  This change was an attempt to fix it.  It turned out that I needed the following similar change in putobject.cpp:

```
            if (current_state == parsing_state::parsing_done) {                                                                                                                                                                                                                                                      
                // this is to force that these are destructed in the correct order                                                                                                                                                                                                                                   
                auto conn_ptr = conn;                                                                                                                                                                                                                                                                                
                auto transport_ptr = tp;                                                                                                                                                                                                                                                                             
                auto dstream_ptr = d;                                                                                                                                                                                                                                                                                
                                                                                                                                                                                                                                                                                                                     
                if (ofs->is_open()) {                                                                                                                                                                                                                                                                                
                    ofs->close();                                                                                                                                                                                                                                                                                    
                }                                                                                                                                                                                                                                                                                                    
                if (!keep_dstream_open_flag && d->is_open()) {                                                                                                                                                                                                                                                       
                    logging::trace("{}:{} Closing iRODS data object.", __func__, __LINE__);                                                                                                                                                                                                                          
                    d->close();                                                                                                                                                                                                                                                                                      
                }     
```
Once I made that change the occasional errors went away.  I left the similar code in completemultipartupload.cpp just to be safe.

The idea is that if I have a local copy of that pointer it won't be deleted when the `std::tuple` is destructed after `erase` and will instead be deleted in the reverse order of how these pointer copies are created when they go out of scope.

Why that is required when `close` has already been called is a mystery to me but it definitely seems to be required.  I am not familiar with the stream code to know exactly why this is the case.

Note that `std::tuple` does not state the order of destruction of its elements.  I tested this with clang and g++ and got difference results.  So I wanted to be able to control it thus I created local copies of these pointers to make sure these objects stay open and are deleted in the right order.

Reply:
Oh, I see. So having references to the shared pointers means that the shared pointers in the tuple are not the last references. Then the tuple gets destructed and these are the last remaining references, which means that they are destructed in reverse when leaving this scope (rather than leaving it up to "chance" in the tuple's destructing). That makes sense. Thanks